### PR TITLE
Multithreaded local ads and stats

### PR DESCRIPTION
**Pull request by rokuz**
https://jira.mail.ru/browse/MAPSME-7008

### CODE REVIEW

File: local_ads/statistics.cpp [hunk 1]
```diff
@@ -335,54 +364,65 @@ void Statistics::ProcessEvents(std::list<Event> & events)
...
+  auto const connectionStatus = GetPlatform().ConnectionStatus();
+  if (connectionStatus == Platform::EConnectionType::CONNECTION_WIFI)
   {
-    std::string const url = MakeRemoteURL(m_userId, it->first.first, it->first.second);
-    if (url.empty())
-      return;
-
-    std::list<Event> events = ReadEvents(it->second.m_fileName);
-    if (events.empty())
+    for (auto it = m_metadataCache.begin(); it != m_metadataCache.end();)
```
Comment (darina) [inline local_ads/statistics.cpp]: не вижу, где теперь it инкрементируется?

**Comment by Jobayerha** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  zobayer.eecl@gmail.com
    Domain everbrightbd.com has exceeded the max emails per hour (51/50 (102%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o5.sgmail.github.com ([192.254.113.10]:42507)
	by fighter.vivawebhost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89_1)
	(envelope-from <bounces+848413-4c68-jobayer=everbrightbd.com@sgmail.github.com>)
	id 1esojq-0001VG-IO
	for jobayer@everbrightbd.com; Mon, 05 Mar 2018 06:57:06 -0500
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=TAlnwxN1Db6zLbGnetsYXlkm9wA=; b=REKBL4VoFjDifiRV
	Gdc6kzENwQ9U6DnT8RP6RzWlqXKi7BY51v+zaR7Bw00pXMpxNIFhFso6XWEmSpTZ
	vMkmX8jP1nc+3ujfZkLcNN4oSX1hdLJTmdfFKRy3XbJSOI06lRToBYwLxHP1qUUU
	Ae/O/bPtH5WVLH/5bQHhVmwAwMQ=
Received: by filter0478p1las1.sendgrid.net with SMTP id filter0478p1las1-26003-5A9D306A-5
        2018-03-05 11:56:26.175194175 +0000 UTC
Received: from github-smtp2b-ext-cp1-prd.iad.github.net (github-smtp2b-ext-cp1-prd.iad.github.net [192.30.253.17])
	by ismtpd0010p1iad1.sendgrid.net (SG) with ESMTP id K1sYy3E-SIyDApWhr40C9w
	for <jobayer@everbrightbd.com>; Mon, 05 Mar 2018 11:56:26.131 +0000 (UTC)
Date: Mon, 05 Mar 2018 11:56:26 +0000 (UTC)

**Comment by Jobayerha** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  zobayer.eecl@gmail.com
    Domain everbrightbd.com has exceeded the max emails per hour (52/50 (104%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 25758 characters long; only the first
------ 5000 or so are included here.
Received: from o7.sgmail.github.com ([167.89.101.198]:37819)
	by fighter.vivawebhost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89_1)
	(envelope-from <bounces+848413-4c68-jobayer=everbrightbd.com@sgmail.github.com>)
	id 1esoki-0001cJ-Ff
	for jobayer@everbrightbd.com; Mon, 05 Mar 2018 06:58:00 -0500
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=7xENOwD2lO/WMqyL6TLjOFQBtf8=; b=VT/BcJaRZxbfPhhv
	RG72ENXmRwqI4bqw0479ydjgEemvFJfupTFv59hAQDuv59mbjii1hNK7qBN2SMVl
	ENJAJveU1GUL030FTFDKyiX8FnIS7vtdWBdfsfOX9UqZ9qGAA3vc4PD6+yCAli6E
	IhltnFwer3w7nU1E7qmg38vdVQg=
Received: by filter0234p1iad2.sendgrid.net with SMTP id filter0234p1iad2-4971-5A9D309F-E
        2018-03-05 11:57:19.509894837 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0007p1iad1.sendgrid.net (SG) with ESMTP id 9Ks4frSeS0WXLXX_OxiGAA
	for <jobayer@everbrightbd.com>; Mon, 05 Mar 2018 11:57:19.210 +0000 (UTC)
Date: Mon, 05 Mar 2018 11:57:19 +0000 (UTC)

**Comment by Jobayerha** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  zobayer.eecl@gmail.com
    Domain everbrightbd.com has exceeded the max emails per hour (53/50 (106%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 29690 characters long; only the first
------ 5000 or so are included here.
Received: from github-smtp2-ext6.iad.github.net ([192.30.252.197]:55348 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by fighter.vivawebhost.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89_1)
	(envelope-from <noreply@github.com>)
	id 1esola-0001jT-SW
	for jobayer@everbrightbd.com; Mon, 05 Mar 2018 06:58:55 -0500
Date: Mon, 05 Mar 2018 03:58:14 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1520251094;
	bh=8WEs7L/CBkOichHJ2m+K7YKUswSQD7+QX0jQIgYH9sk=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=pWFhELSSfNvYDTiP32/KQYF5zK9U8RWd57wpr/E3TYnea00tbDjF9zxrRPSDC0r2Y
	 ZfEC4/WFyx9pVIh21igIVYWJ5iM8iG+i7RqBkBgOdxpt/SoU5Y09WcppD1tAshpspT
	 dVHQQ7XmZj2jxJXi4/6utxYT+sS7kGUn1/c8NpVQ=

**Comment by Jobayerha** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  zobayer.eecl@gmail.com
    Domain everbrightbd.com has exceeded the max emails per hour (54/50 (108%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 26631 characters long; only the first
------ 5000 or so are included here.
Received: from o6.sgmail.github.com ([192.254.113.101]:47313)
	by fighter.vivawebhost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89_1)
	(envelope-from <bounces+848413-4c68-jobayer=everbrightbd.com@sgmail.github.com>)
	id 1esomP-0001q8-EO
	for jobayer@everbrightbd.com; Mon, 05 Mar 2018 06:59:45 -0500
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=mLBr1ZkIaqcSAUdRBupIpht+xOA=; b=bhGlLVBjk7PMz4tk
	ZsokLQZBm92/kXsH2xGYUvCnU8a9hBa1dy4r1L9yGz9vRX3CI2zHvsgJXUk2FTOL
	kT71sgnZXFzdzrRg6X0gjxhkvieH+R0Pf4Z4A9P0rvEcRjroM1TuoI2hghYapW1I
	6EVl902IhjMsSxAYz5FmHqGRBXw=
Received: by filter0219p1las1.sendgrid.net with SMTP id filter0219p1las1-11111-5A9D3109-4
        2018-03-05 11:59:05.19381128 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0004p1iad1.sendgrid.net (SG) with ESMTP id 8O3SihnfTcaDBtveJ9oWCg
	for <jobayer@everbrightbd.com>; Mon, 05 Mar 2018 11:59:04.947 +0000 (UTC)
Date: Mon, 05 Mar 2018 11:59:05 +0000 (UTC)

File: local_ads/statistics.cpp (hunk 1 above)
Comment (milchakov) [reply local_ads/statistics.cpp]: 😜

File: local_ads/statistics.cpp [hunk 2]
```diff
@@ -335,54 +364,65 @@ void Statistics::ProcessEvents(std::list<Event> & events)
...
-                                     ? m_serverSerializer(events, m_userId, contentType, contentEncoding)
-                                     : SerializeForServer(events);
-    ASSERT(!bytes.empty(), ());
-
-    platform::HttpClient request(url);
-    request.SetTimeout(5);    // timeout in seconds
+  }
+  
+  // Send every |kSendingTimeout|.
+  GetPlatform().RunDelayedTask(Platform::Thread::File, kSendingTimeout, [this]
```
Comment (darina) [inline local_ads/statistics.cpp]: kSendingTimeoutInSec

File: local_ads/statistics.cpp [hunk 3]
```diff
@@ -335,54 +364,65 @@ void Statistics::ProcessEvents(std::list<Event> & events)
...
-                        contentEncoding);
-    if (request.RunHttpRequest() && request.ErrorCode() == 200)
-    {
-      FileWriter::DeleteFileX(it->second.m_fileName);
-      it = m_metadataCache.erase(it);
-    }
-    else
+  request.SetBodyData(std::string(bytes.begin(), bytes.end()), contentType, "POST",
+                      contentEncoding);
+  if (request.RunHttpRequest() && request.ErrorCode() == 200)
```
Comment (VladiMihaylenko) [inline local_ads/statistics.cpp]: используй, пожалуйста,   ```bool RunHttpRequest(string & response, SuccessChecker checker = nullptr)```

File: local_ads/statistics.cpp (hunk 1 above)
Comment (rokuz) [reply local_ads/statistics.cpp]: спасибо!

File: local_ads/statistics.cpp (hunk 2 above)
Comment (rokuz) [reply local_ads/statistics.cpp]: Тут не фиксирован временной тип

Comment (darina) [reply local_ads/statistics.cpp]: ок

File: local_ads/statistics.cpp (hunk 3 above)
Comment (rokuz) [reply local_ads/statistics.cpp]: Обсудили, оставим как есть

File: map/local_ads_manager.cpp [hunk 4]
```diff
@@ -395,58 +354,66 @@ bool LocalAdsManager::DownloadCampaign(MwmSet::MwmId const & mwmId, std::vector<
...
-          CreateLocalAdsMarks(m_bmManager, campaignData);
+          auto campaignData = ParseCampaign(std::move(info.m_data), mwm, info.m_created);
+          if (!campaignData.empty())
+          {
+            UpdateFeaturesCache(ReadCampaignFeatures(m_readFeaturesFn, campaignData));
+            CreateLocalAdsMarks(m_bmManager, campaignData);
+          }
         }
+        
+        std::lock_guard<std::mutex> lock(m_campaignsMutex);
```
Comment (milchakov) [inline map/local_ads_manager.cpp]: сделай, пожалуйста, скоуп для мьютекса

File: map/local_ads_manager.cpp [hunk 5]
```diff
@@ -395,58 +354,66 @@ bool LocalAdsManager::DownloadCampaign(MwmSet::MwmId const & mwmId, std::vector<
...
+      auto const & type =  request.second;
+      
+      if (!mwm.IsAlive())
         continue;
-
-      // Parse data and recreate marks.
-      ClearLocalAdsForMwm(mwm);
-      if (!info.m_data.empty())
+      
+      std::string const countryName = mwm.GetInfo()->GetCountryName();
```
Comment (milchakov) [inline map/local_ads_manager.cpp]: `std::string const &`?

File: map/local_ads_manager.cpp (hunk 4 above)
Comment (milchakov) [reply map/local_ads_manager.cpp]: обсудили, решили оставить так как есть

**Pull request merged**